sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala

```diff
@@ -89,6 +94,23 @@ class UserDefinedTypeSuite extends QueryTest with SharedSQLContext with ParquetT
     assert(featuresArrays.contains(new MyDenseVector(Array(0.2, 2.0))))
   }
 
+  test("user type with ScalaReflection") {
+    val points = Seq(
+      MyLabeledPoint(1.0, new MyDenseVector(Array(0.1, 1.0))),
+      MyLabeledPoint(0.0, new MyDenseVector(Array(0.2, 2.0))))
+
+    val schema = ScalaReflection.schemaFor[MyLabeledPoint].dataType.asInstanceOf[StructType]
+    val attributeSeq = schema.toAttributes
+
+    val pointEncoder = encoderFor[MyLabeledPoint]
+    val unsafeRows = points.map(pointEncoder.toRow(_).copy())
```

Comment:
I added the test now. By doing this, I also found that we need to add UserDefinedType to Cast to make it work.